src/Data/Color/Color.php

```diff
@@ -0,0 +1,178 @@
+<?php
+/* Copyright (c) 2017 Nils Haagen <nils.haagen@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\Data\Color;
+use ILIAS\Validation as VAL;
+use ILIAS\Data as D;
+/**
+ * Color expresses a certain color by giving the mixing ratio
+ * in the RGB color space.
+ *
+ * @author Nils Haagen <nils.haagen@concepts-and-training.de>
+ */
+class Color {
+
+	/**
+	 * @var string
+	 */
+	protected $value;
+
+	public function __construct($value) {
+		$this->value = $this->normalize($value);
+	}
+
+	/**
+	 * Return the hex-value of color
+	 *
+	 * @return string
+	 */
+	public function value() {
+		return $this->value;
+	}
+
+	/**
+	 * Return string with rgb-notation
+	 *
+	 * @return string
+	 */
+	public function rgbstring() {
+		return 'rgb('
+			.implode($this->rgb(), ', ')
+			.')';
+	}
+
+	/**
+	 * Return a list with rgb-values of color
+	 *
+	 * @return <int>
+	 */
+	public function rgb() {
+		$col = $this->trimhash($this->value);
+		$chunks = str_split($col, 2);
+		return array_map('hexdec', $chunks);
+	}
+
+	/**
+	* Based on https://de.wikipedia.org/wiki/Luminanz
+	* this funktion decides if the color can be considered "dark".
+	* With a dark background, i.e., a lighter (white) color should be used
+	* for the foreground.
+	*/
+	public function isDark() {
+		$rgb = $this->rgb();
+		$r = $rgb[0];
+		$g = $rgb[1];
+		$b = $rgb[2];
+		$sum = 0.299 * $r + 0.587 * $g + 0.114 * $b;
+		if($sum < 128){
+			return true;
+		}
+		return false;
+	}
+
+
+	/**
+	 * A value is valid, if it
+	 * -starts with '#'
+	 * -is of length 3(4) or 6(7)
+	 * OR
+	 * -is a list of exactly three values
+	 * -each value is between 0 and 255
+	 *
+	 * The datatype stores the value in hex longhand notation.
+	 * Normalization also includes validation.
+	 *
+	 * @throws Exception
+	 * @return string
+	 */
+	private function normalize($value) {
```

Comment:
Move all this specific cases of how a value could be inputted to a new `ColorFactory`. Give them names (like "fromHex", "fromRGB", ...) and split them apart. Make central `build`-method, that decides which `from`-method to use. I.e.: refactor this =)